external/fv3fit/fv3fit/reservoir/autoencoder.py

```diff
@@ -26,20 +41,195 @@ def encode(self, x: Union[np.ndarray, tf.Tensor]) -> np.ndarray:
+                f"input variables {self.input_variables}."
+            )
+
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables)
+
+
+def _get_Xy_dataset(data, variables):
+    concat_data = data.map(partial(concat_variables_along_feature_dim, variables))
+    return tf.data.Dataset.zip((concat_data, concat_data))
+
+
+def _build_autoencoder(
+    sample_data: tf.Tensor,
+    input_size: int,
+    latent_dim_size: int,
+    units: int,
+    n_dense_layers: int,
+) -> Tuple[tf.keras.Model, tf.keras.Model]:
+    encoder_layers = [
+        tf.keras.layers.Input(shape=(input_size), name="input"),
+    ]
+    norm = StandardNormLayer(name="standard_normalize")
+    norm.fit(sample_data)
+
+    encoder_layers.append(norm)
+    for i in range(n_dense_layers):
+        encoder_layers.append(
+            tf.keras.layers.Dense(units, activation="relu", name=f"encoder_{i}")
+        )
+    encoder_layers.append(tf.keras.layers.Dense(latent_dim_size, activation="relu"))
+    encoder = tf.keras.Sequential(encoder_layers)
+
+    decoder_layers = [
+        tf.keras.layers.Input(shape=(latent_dim_size), name="decoder_input"),
+    ]
+    for i in range(n_dense_layers):
+        decoder_layers.append(
+            tf.keras.layers.Dense(units, activation="relu", name=f"decoder_{i}")
+        )
+    decoder_layers.append(
+        tf.keras.layers.Dense(input_size, activation="linear", name="decoded_output")
+    )
+    denorm = StandardDenormLayer(name="standard_denormalize")
+    denorm.fit(sample_data)
+    decoder_layers.append(denorm)
+
+    decoder = tf.keras.Sequential(decoder_layers)
+    return encoder, decoder
+
+
+@register_training_function("dense_autoencoder", DenseAutoencoderHyperparameters)
+def train_dense_autoencoder(
+    hyperparameters: DenseAutoencoderHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+) -> AutoencoderPredictor:
+    norm_batch = next(iter(train_batches))
+    norm_batch_concat = concat_variables_along_feature_dim(
+        variables=hyperparameters.input_variables, variable_tensors=norm_batch
+    )
+    stddev = np.std(norm_batch_concat, axis=0)
+
+    callbacks = [callback.instance for callback in hyperparameters.callbacks]
+    train_Xy = _get_Xy_dataset(train_batches, hyperparameters.input_variables)
+    if validation_batches is not None:
+        val_Xy = _get_Xy_dataset(validation_batches, hyperparameters.input_variables)
+    else:
+        val_Xy = None
+
+    n_features = norm_batch_concat.shape[-1]
+    encoder, decoder = _build_autoencoder(
+        sample_data=norm_batch_concat,
+        input_size=n_features,
+        latent_dim_size=hyperparameters.latent_dim_size,
+        units=hyperparameters.units,
+        n_dense_layers=hyperparameters.n_dense_layers,
+    )
+
+    train_model = Autoencoder(encoder=encoder, decoder=decoder)
+    predict_model = Autoencoder(encoder=encoder, decoder=decoder)
```

Comment:
Suggestion (not required, but would improve maintainability): If you combined these into a single ModelBuilder (a callable function that takes sample data to be used for normalization/sizes as input and returns a train and predict model as output), you could use the `fv3fit.keras._models.dense.train_column_model` to avoid re-writing data processing and configuration management logic. You can see the `train_dense_model` function for an example of this.

Reply:
Updated to use train_column_model and a subclass of PureKerasModel (needed a different dump/load to avoid weird errors when calling .predict on loaded Autoencoder)